plugins/metric_links/src/Metric.php

```diff
@@ -163,32 +149,32 @@ public function scan($uri, \DOMXPath $xpath, $depth, Page $page, Metrics $contex
      * This method will find broken links and mark a page appropriately
      * 
      * @param Page $page the page to mark
-     * @param array $links an array of the links found on the page
      */
-    public function markPage(Page $page, $links)
+    public function markPage(Page $page)
     {
-        $statuses = $this->getStatuses(array_unique($links));
-        $counts = array_count_values($links);
-
-        foreach ($statuses as $url=>$status) {
-            if (!$this->isError($status)) {
+        $links = $page->getLinks();
+        
+        foreach ($links as $link) {
+            if (!$this->isError($link)) {
                 //don't mark it...
                 continue;
             }
 
-            $machine_name = $this->getMachineNameForStatus($status);
+            $machine_name = $this->getMachineNameForStatus($link);
```

Comment:
I like the abstraction of `Metric` logic from the `Link` logic.  `links` are used in multiple places, and they just happened to be reused by the `link metric`, so I don't think a `Link` should know anything about a `Mark`.

A lot of these methods are indeed used by foreign objects.  Some are public just to make testing easier, which isn't a great reason.